[cdac] start of RuntimeTypeSystem contract; implement GetMethodTableData SOS method

lambdageek opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/99302

Also updates the existing DAC so that it does not store an `EEClass*` in `DacpMethodTableData::Class`, but instead stores a pointer to the canonical method table.  Correspondingly `GetMethodTableForEEClass` now expects a canonical MethodTable.

Also renamed `MethodTable::ContainsPointers` to `ContainsGCPointers`, (and corresponding flag and setter method); also updated gcenv for NativeAOT with the same renaming.

Also changed the existing DAC so that `DacpMethodTableData::wNumVirtuals` and `DacpMethodTableData::wNumVtableSlots` are always returned as 0.  The cDAC Metadata contract does not expose the underlying info

verified that it works for at least one MethodTable by running some manual WinDbgX test

TODO:
- [x] write the contract in docs/
- [x] unit tests

lambdageek committed:
WIP: Metadata contract

elinor-fung committed:
Implement GetThreadStoreData in cDAC

elinor-fung committed:
Add placeholder for getting thread data

elinor-fung committed:
Apply suggestions from code review

lambdageek committed:
Merge remote-tracking branch 'elinor-fung/cdac-threadstore' into cdac-wip

lambdageek committed:
fix build

lambdageek committed:
WIP: ValidateMethodTable

elinor-fung committed:
DataCache.GetOrAdd

lambdageek committed:
wip

lambdageek committed:
Merge remote-tracking branch 'elinor-fung/cdac-threadstore' into cdac-wip

lambdageek committed:
checkpoint: ValidateWithPossibleAV

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

lambdageek committed:
checkpoint EEClass from MethodTable

lambdageek committed:
checkpoint: ValidateMethodTablePointer

lambdageek committed:
cp: delegate from legacy dac

lambdageek committed:
add Metadata to runtime contract descriptor

lambdageek committed:
checkpoint: more MethodTable fields

lambdageek committed:
checkpoint GetMethodTableData implemented

lambdageek committed:
checkpoint: same answer as legacy dac

lambdageek commented on src/native/managed/cdacreader/src/Contracts/Metadata.cs:
```diff
@@ -0,0 +1,386 @@
+using System;
+using System.Reflection.Metadata.Ecma335;
+using UntrustedMethodTable = Microsoft.Diagnostics.DataContractReader.Contracts.UntrustedMethodTable_1;
+using MethodTable = Microsoft.Diagnostics.DataContractReader.Contracts.MethodTable_1;
```
I'm trying something here that I'm not completely sure about - the Metadata contract needs to create and consume these MethodTable values, but it's quite likely that their internals would be different once the Metadata contract evolves.  But initially it's annoying to write `MethodTable_1` everywhere.

An alternative is to make a completely opaque `MethodTableHandle` that forms the contract's interface and it will likely carry a pointer back to the contract and the contract will have some kind of dictionary to retrieve the actual interesting method table data.

lambdageek replied:
I decided I don't like this.  I'm going to make `MethodTable` completely opaque for the public contract surface and put all the query methods on `Metadata`

lambdageek commented on src/native/managed/cdacreader/src/Contracts/Metadata.cs:
```diff
@@ -0,0 +1,386 @@
+
+    public virtual bool IsString(in MethodTable methodTable) => throw new NotImplementedException();
+    public virtual bool ContainsPointers(in MethodTable methodTable) => throw new NotImplementedException();
+    public virtual bool IsDynamicStatics(in MethodTable methodTable) => throw new NotImplementedException();
```
I don't understand yet what this flag means or why SOS cares about it

am11 replied:
Is it so that all statics are not considered dynamic (since https://github.com/dotnet/runtime/pull/99183 `All statics are now considered "dynamic" statics.`). If it's representing the same data, then it can be hardwired return `true`?

lambdageek replied:
yea things will be simpler in this respect in 99183.  I need to update to sync things up

lambdageek commented:
Need to reconcile with https://github.com/dotnet/runtime/pull/99183 - I was on an older branch

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

lambdageek committed:
new flags for statics

jkotas commented on src/native/managed/cdacreader/src/Contracts/Metadata.MethodTableFlags.cs:
```diff
@@ -0,0 +1,190 @@
+
+        UNUSED_ComponentSize_1 = 0x00000001,
+        // GC depends on this bit
+        HasCriticalFinalizer = 0x00000002, // finalizer must be run on Appdomain Unload
```
Similar to my comment from other PR. Do all these bits need to be part of the diagnostic contract?

lambdageek replied:
I need to do a bit of refactoring.

There are two sides to the contract:
* the information that cdac consumes from the runtime.  I think probably we don't need most of the bits from the method table flags.  I'll remove the ones we don't need
* the public contract surface.  Right now `MethodTable_1` is leaky because it does actually expose the flags dwords completely.   I think I don't actually like having `MethodTable_1` having any public properties - all public interaction with it should be through a `Metadata`  contract subclass which should only expose the logical interpretation of the flags bits ("is string", "is generic instance", etc)

lambdageek committed:
fix GCC build

lambdageek commented:
I wonder if we really need to expose EEClass from DacpMethodTableData - it looks like it's only used in output (for example here: https://github.com/dotnet/diagnostics/blob/761d22e2673aa5d3859a142cbe24e348ea93a96c/src/SOS/Strike/strike.cpp#L1080)
I wonder if I can either always return 0, or maybe just return the address of the MethodTable itself (or the canonical method table or something)

lambdageek committed:
WIP: opaque MethodTableHandle

mikem8361 commented:
It is used more than just display in SOS.  It is the !dumpclass parameter to dump more info about a class.  There is a DML link [here](https://github.com/dotnet/diagnostics/blob/761d22e2673aa5d3859a142cbe24e348ea93a96c/src/SOS/Strike/util.cpp#L4539) (only works under windbg) that links the EEClass address to the !dumpclass command.

We also need to check to make sure [CLRMD](https://github.com/microsoft/clrmd/tree/main) doesn't use that field too.  It looks like it currently doesn't (it is returned [here](https://github.com/microsoft/clrmd/blob/066e952ad7a85b7ba495b1209e4ebb93b15bed97/src/Microsoft.Diagnostics.Runtime/DacInterface/SosDac.cs#L436)) but the EEClass field isn't used.  There is the SosDac.cs function [GetMethodTableByEEClass](https://github.com/microsoft/clrmd/blob/066e952ad7a85b7ba495b1209e4ebb93b15bed97/src/Microsoft.Diagnostics.Runtime/DacInterface/SosDac.cs#L528) that takes a EEClass address but there are no callers.

My point is that SOS isn't the only one using the DAC APIs.

jkotas commented:
We can make EEClass = canonical MethodTable in all these places.

lambdageek committed:
Add contract accessors for MethodTableHandle

lambdageek committed:
fixup

lambdageek committed:
simplify FreeObjectMethodTable handling

lambdageek committed:
cleanup

lambdageek committed:
fixup

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

lambdageek committed:
[dac] Return canonical MethodTable instead of EEClass

Instead of storing the EEClass pointerin
DacpMethodTableData, store the canonical method table instead.

Correspondingly, update GetMethodTableForEEClass to expect a canonical method table pointer instead of an EEClass

Also update cDAC to do likewise

lambdageek committed:
Delete unreferenced MethodTable flags

lambdageek committed:
add Metadata contract doc; fixups

lambdageek commented on docs/design/datacontracts/Metadata.md:
````diff
@@ -0,0 +1,286 @@
+``` csharp
+internal partial struct Metadata_1
+{
+    [Flags]
````
> I'd prefer to avoid these types.

Avoid them how?  This is how CoreCLR stores the bits.  this is an `internal` part of Metadata version 1. It doesn't leak out into the public `IMetadata` contract surface

> > We should do a bit better here. Is this the data descriptors or th contract? I'm struggling a bit with terms here.

There are three terms, in my mind:

* data descriptor: describes the layout of structures for a particular (version of a) runtime.  A contract implementation (`Metadata_1`) has a dependency on some data descriptors (that is - a runtime that provides `Metadata` version 1 is expected to provide a descriptors named `MethodTable` and `EEClass`)
* contract backend: the interpretation of the data in the target memory for a particular version of a contract. This is the `private` and `internal` fields/methods/structs of a particular contract (`Metadata_1` in our case)
* contract frontend: this is the `public` API surface provided by a contract (for example the `IMetadata` interface) - this is how you interact with the contract from the outside, for example to implement the SOS DAC APIs. 

Examples:

1. If CoreCLR changes where the `dwFlags` and `dwFlags2` fields are inside `MethodTable`, that just requires a change of the data descriptor.  no version bumps are necessary. an updated data descriptor embedded in the runtime will always be in sync.
2. If CoreCLR moves some flags or fields from `MethodTable` to `EEClass` but conceptually they represent the same logic, that will require a contract version bump and some of the `internal` members of `Metadata_vNext` will need to change, but the public contract API will stay the same.
3. If .NET 17 adds a new kind of type (for example some kind of zero-sized types like in Rust, that only exist for type computation) and they're relevant for diagnostic APIs, we would add some new accessor/query methods to `IMetadata` (with some kind of default behavior for older runtimes)

lambdageek commented on docs/design/datacontracts/Metadata.md:
```diff
@@ -0,0 +1,286 @@
+    }
+}
+
+internal struct EEClass_1
```
it's an implementation detail of `Metadata` version 1. it's not a public contract API.  The only public part is a `MethodTableHandle` which is completely opaque

lambdageek commented on src/native/managed/cdacreader/src/Legacy/ISOSDacInterface.cs:
```diff
@@ -42,6 +42,25 @@ internal struct DacpThreadData
+{
+    public int bIsFree; // everything else is NULL if this is true.
+    public ulong module;
+    public ulong @class;
```
This is a transcription of the DAC COM .idl - I wasn't sure it's ok to change the field names

AaronRobinsonMSFT replied:
You can change these field names. No one should be using TLB referencing for the interop layout - that is where the names are important.

lambdageek commented on src/native/managed/cdacreader/src/Legacy/SOSDacImpl.cs:
```diff
@@ -81,9 +82,63 @@ public int GetBreakingChangeVersion()
+            DacpMethodTableData result = default;
+            result.baseSize = contract.GetBaseSize(methodTable);
+            if (contract.IsString(methodTable))
+                result.baseSize -= 2 /*sizeof(WCHAR) */;
```
No idea.

@mikem8361 why does the SOS DAC do this:

https://github.com/dotnet/runtime/blob/24b59d85541c72fd55b8d5d0bc9adb84619d6c4f/src/coreclr/debug/daccess/request.cpp#L1799-L1800

am11 replied:
Maybe better:
```suggestion
                result.baseSize -= sizeof(char);
```

jkotas replied:
The DAC APIs are stuck in time. As you know there are many places where the DAC implementation returns dummy values or constants to keep the DAC API "working". This is another one of those cases. The BaseSize implementation in the runtime changed. but something that consumed DAC API was expecting the old BaseSize value for string and this was added to keep things "working".

For reference, this is the checkin description for the change that introduced this adjustment:

```
	2008/09/25

	Title: New implementation of StringBuilder and improvements in String class
	
	Issue description:
	  1. The existing implementation of StringBuilder uses a String object as
	  its internal buffer. Every time we are short of space, we double the
	  String size and very soon we find ourselves into the large object heap.
	  Since Gen2 collections do not happen as often as Gen0 and Gen1
	  collections, these objects (even if no longer used) remain allocated
	  2. Secondly, this dependency on String necessitates the need to
	  differentiate between arrayLength and stringLength. With the change in 1
	  above, arrayLength is always 1 more than stringLength (the
	  null-terminator) and we can get rid of arrayLength and derive that from
	  stringLength based on need.
	
	Change description:
	  1. This new implementation implements StringBuilder as a linked list of
	  StringBuilder (called chunks). The buffer used internally by the
	  StringBuilder is a char array which is doubled until the chunk size
	  reaches a cap of 8K after which we allocate a new chunk. Almost all the
	  implementation in StringBuilder.cs has been done by Vance.
	  a. StringBuilder.Append faster by 25% on an average
	  b. StringBuilder.ToString is as performant as before
	  2. Remove the bufferLength from the String class. 4 bytes saved per
	  String (except empty Strings where we do padding). So working set gains
	  Different CR'ers have reviewed different sections of the code. Jan and
	  Brian sort of looked at the overall change
```

I hope we will avoid shenanigans like this one going forward by having explicit versioning scheme.

AaronRobinsonMSFT replied:
Thanks @jkotas. Jeremy and I have been playing a lot of games recently with these sorts of semantics that we respect but are historical artifacts. We often label them with a comment such as below to indicate this is present due to compat reasons.

> `// [Compat] <MINOR DETAIL ON WHAT WE FOUND>`

jkotas replied:
This adjustment is complementary to the inverse adjustment done in the SOS here: https://github.com/dotnet/diagnostics/blob/464e06c9a8602dff44ff5c59e9a0c3777f68e41c/src/SOS/Strike/eeheap.cpp#L447-L450 . In the ideal world, we would delete both.

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

lambdageek committed:
rename DacpMethodTableData:klass field

lambdageek committed:
document GetMethodTableData string baseSize adjustment

lambdageek committed:
Apply suggestions from code review

Co-Authored-By: Aaron Robinson <arobins@microsoft.com>

lambdageek commented:
Question:

Looking ahead a bit at `GetMethodTableName`, I see that I'm going to need to work with CoreCLR's `TypeHandle`.  A `TypeHandle` is a single pointer-sized value that is either a `MethodTable*` or a tagged `TypeDesc*`.  Since all raw `MethodTable*` are valid `TypeHandle`s, I wonder if it's better if the contract was in terms of `TypeHandle` and we wouldn't have to expose both concepts.

AaronRobinsonMSFT commented:
> Question:
> 
> Looking ahead a bit at `GetMethodTableName`, I see that I'm going to need to work with CoreCLR's `TypeHandle`. A `TypeHandle` is a single pointer-sized value that is either a `MethodTable*` or a tagged `TypeDesc*`. Since all raw `MethodTable*` are valid `TypeHandle`s, I wonder if it's better if the contract was in terms of `TypeHandle` and we wouldn't have to expose both concepts.

I agree with your general sentiment that exposing both is less than ideal and we should abstract it away and consider it an implementation detail. Not sure if the current `TypeHandle` is the best abstraction/name, so I'd like to see/hear what alternatives we can come up with.

mikem8361 commented:
>> @mikem8361<https://github.com/mikem8361> why does the SOS DAC do this:

>>

>> https://github.com/dotnet/runtime/blob/24b59d85541c72fd55b8d5d0bc9adb84619d6c4f/src/coreclr/debug/daccess/request.cpp#L1799-L1800

I have no idea either, but I think Jan already answered this question.

lambdageek commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.cs:
```diff
@@ -0,0 +1,385 @@
+            Metadata_1.MethodTableFlags flags = new Metadata_1.MethodTableFlags
+            {
+                DwFlags = _target.Read<uint>(Address + (ulong)_type.Fields[nameof(Metadata_1.MethodTableFlags.DwFlags)].Offset),
+                DwFlags2 = _target.Read<uint>(Address + (ulong)_type.Fields[nameof(Metadata_1.MethodTableFlags.DwFlags)].Offset),
```
yikes! I think I need a more complicated test

lambdageek commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.cs:
```diff
@@ -0,0 +1,385 @@
+    internal ushort NumMethods { get; }
+    internal ushort NumNonVirtualSlots { get; }
+    internal uint TypeDefTypeAttributes { get; }
+    internal EEClass_1(Data.EEClass eeClassData)
```
I think this is leftovers from before I changed the public surface of the contract to be in terms of `MethodTableHandle`.  At that point `EEClass_1` leaked out since `GetClassData` was public

jkotas commented on docs/design/datacontracts/Metadata.md:
```diff
@@ -0,0 +1,287 @@
+    // True if the MethodTable is the sentinel value associated with unallocated space in the managed heap
+    public virtual bool IsFreeObjectMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual bool IsString(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // True if the MethodTable represents a type that contains managed references
+    public virtual bool ContainsPointers(MethodTableHandle methodTable) => throw new NotImplementedException();
```
This flag is only set when the type contains objrefs. It is not set when the type only contains byrefs that are also managed references. `ContainsManagedReferences` is not the right name to use here.

I agree that `Pointers` is not the best name, but it is used in hundreds of places throughout the codebase with this meaning. You can call it `ContainsGCPointers` that we use in bunch of places. I would rename the flag and the method in the runtime implementation too if you go with this suggestion.

lambdageek replied:
Renamed everywhere (including GC interface and NativeAOT) to ContainsGCPointers.

jkotas commented on docs/design/datacontracts/Metadata.md:
```diff
@@ -0,0 +1,287 @@
+    public virtual bool IsDynamicStatics(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumMethods(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumInterfaces(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumVirtuals(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumVtableSlots(MethodTableHandle methodTable) => throw new NotImplementedException();
```
I would question the SOS need to print these numbers. These numbers go into the MethodTable data layout details. Are the number of slots actually needed by anything else except some SOS command printing them?

lambdageek replied:
We discussed this a little bit in our standup.  The explanation was that some of these fields, while they're only used for SOS command output are important for a number of reasons:

1. Some 3rd party tooling scrapes SOS output. So eliminating the slot counts from the output entirely or changing the ouptut format too drastically may end up breaking those uses
2. The actual values are useful for engineers working on bringing up the runtime on new platforms

jkotas replied:
I am sorry I was not able to make it to the standup today.

> Some 3rd party tooling scrapes SOS output.

I do not see a problem with that. The screen scrapping done by 3rd party tools should not prevent us from making the SOS better.

> The actual values are useful for engineers working on bringing up the runtime on new platforms

The main value of SOS dump commands - for both platform bring up and general debugging - is in displaying information that is hard (labor intensive) to get manually via regular windbg/lldb commands. Type name is an example of such information.

NumVtableSlots is easy to get using regular windbg/lldb commands, only needed to interpret data layout of the methodtable, and rarely used for low-level debugging. It does not need to be part of Dump... output. I would even say that Dump... would be better without it.. We should be free to delete it to make SOS better. https://github.com/dotnet/coreclr/pull/16560 or https://github.com/dotnet/diagnostics/pull/4751 are examples of changes like that.

AaronRobinsonMSFT replied:
> I do not see a problem with that. The screen scrapping done by 3rd party tools should not prevent us from making the SOS better.

I agree with this in the long term and how we should frame the ultimate outcome of the cDAC workstream. For these initial phases though I tend to consider them more proving this can be done in a seamless and transparent manner. This implies ensuring we maintain the illusion of the static nature should be a priority until we can announce changes in this space and start to innovate. Basically, for now I think we should match precise output unless there is a compelling or hard-blocking reason to deviate from that.

jkotas replied:
> for now I think we should match precise output unless there is a compelling or hard-blocking reason to deviate from that.

Are you saying that we should revert https://github.com/dotnet/runtime/commit/32665b21780fbb54834457d9f1b918b1dc5ea9cf to match existing output    and that we should avoid [cleaning up clutter in the output of SOS commands](https://github.com/dotnet/diagnostics/pull/4751#issuecomment-2184315458)?

It does not sound right to me. I think we should take every opportunity to make small incremental simplifications as we run into them.

AaronRobinsonMSFT replied:
> Are you saying that we should revert https://github.com/dotnet/runtime/commit/32665b21780fbb54834457d9f1b918b1dc5ea9cf to match existing output and that we should avoid https://github.com/dotnet/diagnostics/pull/4751#issuecomment-2184315458? Both of these changes can break screen scraping tools.

As an initial goal, I would've said "yes, we should stick to avoiding breaking changes in this space". However, it seems like that ship has already sailed. I want to innovate in this space a great deal so as an end goal I agree. My concern was and is the burden this can place on automation that relies on scapping. We end up encouraging odd solutions as people try to anticipate or adapt to our "infight work". Since it seems that isn't a current goal, I agree then that we should remove as much cruft as possible.

AaronRobinsonMSFT replied:
Perhaps there should be an SOS query that can help scrapping tooling know what they can expect? Is a new SOS command along the lines of `!sos.version` appropriate in this case? I truly don't know how onerous this is so I could be concerned about something that has limited impact. Perhaps @leculver can shed some light on a version command option, if one doesn't already exist.

lambdageek replied:
Updated the current PR to always set `DacpMethodTableData::wNumVirtuals` and `DacpMethodTableData::wNumVtableSlots` to 0 in both the cDAC and the Legacy DAC.   Updated the Metadata contract to exclude getters for this info.

lambdageek replied:
Related change in dotnet/diagnostics to not print the counts if they're 0 https://github.com/dotnet/diagnostics/pull/4760

I checked and https://github.com/Microsoft/clrmd doesn't use these fields. They only look at `wNumMethods`

leculver replied:
> Perhaps there should be an SOS query that can help scrapping tooling know what they can expect? Is a new SOS command along the lines of `!sos.version` appropriate in this case? I truly don't know how onerous this is so I could be concerned about something that has limited impact. Perhaps @leculver can shed some light on a version command option, if one doesn't already exist.

Sorry I missed this ping.

My guiding principle is:  When possible, and when it doesn't cause too much work, don't regress or take back features that affect debugging.  That includes .Net Core and Desktop Framework (since we can debug 4.5+ desktop framework with current SOS).

That's my only real request.

> Perhaps there should be an SOS query that can help scrapping tooling know what they can expect? Is a new SOS command along the lines of !sos.version appropriate in this case? 

I don't think we need that.  **The output of SOS does not constitute an interface.**  We shouldn't bend over backwards to accommodate screen scraping.  Literally the original reason ClrMD was literally built was to replace screen-scraping SOS.  We don't have all functionality built into ClrMD, but the SOS source is open-source and folks can reimplement functionality from there if they need to.

I haven't fully dug into everything here, but I'm fine with these changes.  I'd honestly be fine if folks wanted to randomize the output of SOS to _intentionally_ break screen scraping.  (Hyperbole.  Obviously, we would never actually do that.)  I owned SOS 15 years ago and built a lot of its functionality.  Screen scraping was the bane of my existence because we couldn't do _anything_ to SOS in the presence of screen scraping.  Just use ClrMD if you need that programmatic access!  No screen scraping!

lambdageek committed:
fix typo

lambdageek committed:
rename flag to ContainsGCPointers

lambdageek committed:
[vm] rename ContainsPointers flag to ContainsGCPointers

also rename getter/setter methods in MethodTable

lambdageek committed:
code style suggestions from code review

lambdageek committed:
BUGFIX: read DwFlags2 from the correct offset

lambdageek committed:
hide utility methods

lambdageek committed:
remove EEClass_1 struct

use Data.EEClass directly in places where we trust an EEClass pointer

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

lambdageek commented on src/coreclr/debug/runtimeinfo/datadescriptor.h:
```diff
@@ -154,6 +154,34 @@ CDAC_TYPE_BEGIN(GCHandle)
+
+CDAC_TYPE_BEGIN(MethodTable)
+CDAC_TYPE_INDETERMINATE(MethodTable)
+CDAC_TYPE_FIELD(MethodTable, /*uint32*/, DwFlags, cdac_offsets<MethodTable>::m_dwFlags)
```
Will rename the cdac_offsets specialization members.  I want to keep `DwFlags`/`DwFlags2` rather than just `Flags`/`Flags2` just because it's at least a little bit easier to search for.

jkotas replied:
Nit: `dw` (DWORD) is ancient left-over from 16-bit Windows. It would be nice to avoid it and other Hungarian notation in the data contract.

If you want something that is easy to search for, you can make it `MTFlags` or something like that.

lambdageek committed:
rename data descriptor members to remove prefixes

lambdageek committed:
cleanup the contract docs

jkotas commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -2063,27 +2112,53 @@ ClrDataAccess::GetMethodTableTransparencyData(CLRDATA_ADDRESS mt, struct DacpMet
+HRESULT
+ClrDataAccess::GetMethodTableForEEClassImpl(CLRDATA_ADDRESS eeClassReallyCanonMT, CLRDATA_ADDRESS *value)
+{
+    PTR_MethodTable pCanonMT = PTR_MethodTable(TO_TADDR(eeClassReallyCanonMT));
```
Could you please change SOS to print parent method table here: https://github.com/dotnet/diagnostics/blob/main/src/SOS/Strike/strike.cpp#L1083C37-L1083C45 , and maybe delete places where it prints EEClass to avoid confusion.

jkotas replied:
There is some discussion about the best way to do it in https://github.com/dotnet/diagnostics/pull/4751#discussion_r1649695256

lambdageek replied:
https://github.com/dotnet/diagnostics/pull/4758

lambdageek committed:
remove hungariant notation prefixes from data descriptors

lambdageek committed:
DAC: always set wNumVirtuals and wNumVtableSlots to 0

This information can be retreived from the MethodTable using normal lldb/windbg primitives and doesn't need to be part of the DAC API contract

jkotas commented on docs/design/datacontracts/Metadata.md:
```diff
@@ -0,0 +1,220 @@
+    public virtual bool IsDynamicStatics(MethodTableHandle methodTable);
+    public virtual ushort GetNumMethods(MethodTableHandle methodTable);
+    public virtual ushort GetNumInterfaces(MethodTableHandle methodTable);
+    public virtual ushort GetNumVirtuals(MethodTableHandle methodTable);
+    public virtual ushort GetNumVtableSlots(MethodTableHandle methodTable);
```
```suggestion
```

jkotas commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.cs:
```diff
@@ -0,0 +1,337 @@
+
+// GC Heap corruption may create situations where a putative pointer to a MethodTable
+// may point to garbage. So this struct represents a MethodTable that we don't necessarily
+// trust to be valid.
```
None of the data in the target process are trusted to be valid, even after they are validated.

I understand what you are trying to say, but "trust" is probably not the best word to describe it since it has security connotations.

Naming idea: `NonValidatedMethodTable`?

lambdageek committed:
Remove NumVirtuals and NumVtableSlots from Metadata.md

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

lambdageek committed:
"untrusted" -> "non-validated"

lambdageek committed:
merge fixup

catch System.Exception not Data.Exception

lambdageek committed:
remove #if 0

lambdageek committed:
cleanup

lambdageek committed:
pull test target helpers out

goal is to be able to use this for testing contracts that depend on some data in the heap

lambdageek committed:
Add one FreeObjectMethodTable unit test

lambdageek committed:
clean up the test helpers a bit

lambdageek committed:
validate that a mock system object is a valid method table

AaronRobinsonMSFT commented on docs/design/datacontracts/Metadata.md:
```diff
@@ -0,0 +1,217 @@
+        public bool IsArray => GetFlag(WFLAGS_HIGH.Category_Array_Mask) == WFLAGS_HIGH.Category_Array;
+
+        public bool IsStringOrArray => HasComponentSize;
+        public ushort RawGetComponentSize() => (ushort)(MTFlags >> 16);
```
What is `MTFlags` in this case? I don't see that in this contract?

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.cs:
```diff
@@ -0,0 +1,333 @@
+    private readonly Target _target;
+    private readonly TargetPointer _freeObjectMethodTablePointer;
+
+    // FIXME: we mutate this dictionary - copies of the Metadata_1 struct share this instance
```
Can we use `ReadOnlyDictionary` instead?

lambdageek replied:
We _want_ to mutate the dictionary: after a putative method table address has been validated, we work with instnaces from the dictionary.

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.cs:
```diff
@@ -0,0 +1,333 @@
+        return true;
+    }
+
+    private bool ValidateWithPossibleAV(NonValidatedMethodTable_1 methodTable)
```
Is this saying it will A/V on the debuggee side or debugger? I assume the debuggee side since we just call it and we can't catch an A/V in managed code any longer. Perhaps it would be more accurate to simple say `ValidateMethodTableUnsafe()` and document the potential consequences instead of encoding that in the name.

AaronRobinsonMSFT replied:
Can I assume this will not be an `AccessViolationException`? If so, then yes please update to `ValidateThrowing()`.

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.cs:
```diff
@@ -0,0 +1,333 @@
+
+    internal TargetPointer FreeObjectMethodTablePointer => _freeObjectMethodTablePointer;
+
+    private NonValidatedMethodTable_1 GetUntrustedMethodTableData(TargetPointer methodTablePointer)
```
What is "Untrusted" in this case?

lambdageek replied:
Missed one "Untrusted" -> "NonValidated" renaming

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Data/EEClass.cs:
```diff
@@ -0,0 +1,23 @@
+
+    public TargetPointer MethodTable { get; init; }
+    public ushort NumMethods { get; init; }
+    public uint AttrClass { get; init; }
```
Is `AttrClass` an ECMA term? I think we should align with "public" API details instead. I believe this data comes from `CorTypeAttr` defined in `corhdr.h`. It would make searching much easier.

lambdageek replied:
It comes from EEClass::m_dwAttrClass.  ECMA-355 just calls them "Type Attributes (II.23.1.15)".  `CorTypeAttr` is good.

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Data/MethodTable.cs:
```diff
@@ -0,0 +1,33 @@
+        NumVirtuals = target.Read<ushort>(address + (ulong)type.Fields[nameof(NumVirtuals)].Offset);
+    }
+
+    public uint MTFlags { get; init; }
```
Is the `MTFlags` name adding clarity here? It is on the `MethodTable` table so I'd imagine `MT` would be implied. Is there some other nuance I am missing?

lambdageek replied:
Just "Flags" was annoying to search for. I'm not strongly attached to the name, though.

AaronRobinsonMSFT replied:
I see. Yes, that makes sense to me.

AaronRobinsonMSFT replied:
> I see. Yes, that makes sense to me.

Feel free to keep this as is.

lambdageek commented on src/native/managed/cdacreader/src/Contracts/Metadata_1.MethodTableFlags.cs:
```diff
@@ -0,0 +1,87 @@
+        public bool IsArray => GetFlag(WFLAGS_HIGH.Category_Array_Mask) == WFLAGS_HIGH.Category_Array;
+
+        public bool IsStringOrArray => HasComponentSize;
+        public ushort RawGetComponentSize() => (ushort)(MTFlags >> MTFlagsComponentSizeShift);
```
"don't bother checking the HasComponentSize bit"

lambdageek replied:
rearranged things a bit.  "ComponentSizeBits" is a private property that just get the raw value from the MTFlags, and `ComponentSize` is the public property that checks the `HasComponentSize` bit before returning something.

lambdageek committed:
code review feedback and more tests:

1. rename AttrClass data descriptor field to CorTypeAttr
2. fixup HasComponentSize / RawGetComponentSize comments and code
3. update "system.object" mock methodtable with more field values
4. update "system.string" mock methodtable with more field values

lambdageek committed:
Update src/coreclr/gc/env/gcenv.object.h

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
Update src/native/managed/cdacreader/src/Contracts/Metadata_1.MethodTableFlags.cs

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
Address code review feedback

lambdageek committed:
move non-validated MethodTable handling to a separate class

lambdageek committed:
clear up ComponentSize contract spec and impl

lambdageek commented:
I was asked to rename the contract from `Metadata` to something else because CoreCLR doesn't really use "metadata" to talk about the execution-time type identity of managed objects.  The obvious alternative name for the contract is `MethodTable`, but I'm open to other options.

jkotas commented:
Naming suggestion: `TypeSystem` or `RuntimeTypeSystem`

AaronRobinsonMSFT commented:
> Naming suggestion: `TypeSystem` or `RuntimeTypeSystem`

I prefer `RuntimeTypeSystem`.

lambdageek committed:
rename Metadata -> RuntimeTypeSystem

lambdageek committed:
add validation failure test; change validation to throw InvalidOperationException

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-wip

davidwrighton commented on src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs:
```diff
@@ -0,0 +1,157 @@
+    private readonly Target _target;
+    private readonly TargetPointer _freeObjectMethodTablePointer;
+
+    // TODO(cdac): we mutate this dictionary - copies of the Metadata_1 struct share this instance.
```
Support for invalidating our view of memory and re-reading to fill in the world should be a pretty soon now project, but I'm ok with having checking in the current state here, and fixing it later when we add that support.

jkotas replied:
```suggestion
    // TODO(cdac): we mutate this dictionary - copies of the RuntimeTypeSystem_1 struct share this instance.
```
?

lambdageek committed:
Update src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

lambdageek committed:
spellcheck

lambdageek committed:
Merge branch 'cdac-wip' of github.com:lambdageek/runtime into cdac-wip

lambdageek committed:
Add a generic instance test

lambdageek committed:
add array instance test

lambdageek commented:
I now have unit tests for MethodTable validation and access for things that look like: `System.Object`, `System.String`, a methodTable with a garbage EEClass, a generic instance (and its generic definition), an array of some valuetype.  This covers a  variety of paths through`ValidateMethodTablePointer`.

Doing more testing is difficult until we have more query methods and data descriptors to mock something more interesting.

I'd like to finish up with this PR and get into expanding the contract for the next SOS API (most likely `GetMethodTableName`)
